File: packages/sdk-react-native/src/client.test.ts
import { describe, it, expect, vi } from "vitest";
import { createQuoteClient, HealthPayQuoteClient } from "./client";
import { ConflictError, NotFoundError, ValidationError } from "./errors";
import type { Quote } from "./types";

const quote = (status: Quote["status"], selectedOptionId: string | null = null): Quote => ({
  id: "11111111-2222-3333-4444-555555555555",
  status,
  serviceType: "lab_investigation",
  providerType: "labs",
  specialty: "labs",
  governorate: "Cairo",
  area: "Nasr City",
  city: null,
  mobile: "+20 100 •••• 567",
  memberNameEn: "Ahmed",
  memberNameAr: "أحمد",
  expiresAt: "2026-06-12T00:00:00.000Z",
  selectedOptionId,
  options: [
    {
      id: "opt-1",
      providerId: "prov-1",
      isAlternative: false,
      providerName: "Alfa Labs",
      providerAddress: "Nasr City",
      serviceDescription: "CBC",
      listPrice: 800,
      discountedPrice: 320,
      discountPct: 60,
      currency: "EGP",
      validityNote: null,
      extraInfo: null,
    },
  ],
});

function jsonResponse(body: unknown, status = 200): Response {
  return new Response(JSON.stringify(body), {
    status,
    headers: { "content-type": "application/json" },
  });
}

describe("HealthPayQuoteClient", () => {
  it("requires baseUrl and token", () => {
    expect(() => new HealthPayQuoteClient({ baseUrl: "", token: "t" })).toThrow();
    expect(() => new HealthPayQuoteClient({ baseUrl: "https://x", token: "" })).toThrow();
  });

  it("GETs the public token endpoint and strips a trailing slash", async () => {
    const fetchMock = vi.fn().mockResolvedValue(jsonResponse(quote("quoted")));
    const hp = createQuoteClient({ baseUrl: "https://x/", token: "tok 1", fetch: fetchMock });
    const q = await hp.get();
    expect(q.status).toBe("quoted");
    expect(fetchMock).toHaveBeenCalledWith(
      "https://x/api/v1/quote/tok%201",
      expect.objectContaining({ method: "GET" }),
    );
  });

  it("confirms a selection", async () => {
    const fetchMock = vi.fn().mockResolvedValue(jsonResponse(quote("confirmed", "opt-1")));
    const hp = createQuoteClient({ baseUrl: "https://x", token: "t", fetch: fetchMock });
    const q = await hp.confirm("opt-1");
    expect(q.selectedOptionId).toBe("opt-1");
    const [, init] = fetchMock.mock.calls[0]!;
    expect(init.method).toBe("POST");
    expect(JSON.parse(init.body)).toEqual({ optionId: "opt-1" });
  });

  it("maps API error codes to typed errors", async () => {
    const fetchMock = vi
      .fn()
      .mockResolvedValueOnce(jsonResponse({ error: { code: "not_found", message: "nope" } }, 404))
      .mockResolvedValueOnce(jsonResponse({ error: { code: "conflict", message: "bad state" } }, 409))
      .mockResolvedValueOnce(jsonResponse({ error: { code: "validation_error", message: "pick one" } }, 422));
    const hp = createQuoteClient({ baseUrl: "https://x", token: "t", fetch: fetchMock });
    await expect(hp.get()).rejects.toBeInstanceOf(NotFoundError);
    await expect(hp.confirm("x")).rejects.toBeInstanceOf(ConflictError);
    await expect(hp.confirm("x")).rejects.toBeInstanceOf(ValidationError);
  });

  it("polls until a terminal status and then stops", async () => {
    vi.useFakeTimers();
    const fetchMock = vi
      .fn()
      .mockResolvedValueOnce(jsonResponse(quote("pending_quote")))
      .mockResolvedValueOnce(jsonResponse(quote("quoted")))
      .mockResolvedValueOnce(jsonResponse(quote("confirmed", "opt-1")));
    const hp = createQuoteClient({ baseUrl: "https://x", token: "t", fetch: fetchMock });

    const seen: string[] = [];
    hp.poll((q) => seen.push(q.status), { intervalMs: 1000 });

    await vi.advanceTimersByTimeAsync(0); // initial tick
    await vi.advanceTimersByTimeAsync(1000);
    await vi.advanceTimersByTimeAsync(1000);
    await vi.advanceTimersByTimeAsync(2000); // would be a 4th tick, but polling stopped

    expect(seen).toEqual(["pending_quote", "quoted", "confirmed"]);
    expect(fetchMock).toHaveBeenCalledTimes(3);
    vi.useRealTimers();
  });
});

File: packages/sdk-flutter/test/client_test.dart
import 'dart:convert';

import 'package:flutter_test/flutter_test.dart';
import 'package:http/http.dart' as http;
import 'package:http/testing.dart';
import 'package:healthpay_quote_sdk/healthpay_quote_sdk.dart';

Map<String, dynamic> quoteJson(String status, {String? selectedOptionId}) => {
      'id': '11111111-2222-3333-4444-555555555555',
      'status': status,
      'serviceType': 'lab_investigation',
      'providerType': 'labs',
      'specialty': 'labs',
      'governorate': 'Cairo',
      'area': 'Nasr City',
      'city': null,
      'mobile': '+20 100 •••• 567',
      'memberNameEn': 'Ahmed',
      'memberNameAr': 'أحمد',
      'expiresAt': '2026-06-12T00:00:00.000Z',
      'selectedOptionId': selectedOptionId,
      'options': [
        {
          'id': 'opt-1',
          'providerId': 'prov-1',
          'isAlternative': false,
          'providerName': 'Alfa Labs',
          'providerAddress': 'Nasr City',
          'serviceDescription': 'CBC',
          'listPrice': 800,
          'discountedPrice': 320,
          'discountPct': 60,
          'currency': 'EGP',
          'validityNote': null,
          'extraInfo': null,
        },
      ],
    };

http.Response ok(Map<String, dynamic> body) =>
    http.Response(jsonEncode(body), 200, headers: {'content-type': 'application/json'});

http.Response err(String code, int status) => http.Response(
      jsonEncode({'error': {'code': code, 'message': code}}),
      status,
      headers: {'content-type': 'application/json'},
    );

void main() {
  test('requires baseUrl and token', () {
    expect(() => HealthPayQuoteClient(baseUrl: '', token: 't'), throwsA(isA<HealthPayError>()));
    expect(() => HealthPayQuoteClient(baseUrl: 'https://x', token: ''), throwsA(isA<HealthPayError>()));
  });

  test('GETs the public token endpoint and strips a trailing slash', () async {
    Uri? seen;
    final mock = MockClient((req) async {
      seen = req.url;
      return ok(quoteJson('quoted'));
    });
    final hp = HealthPayQuoteClient(baseUrl: 'https://x/', token: 'tok 1', httpClient: mock);
    final q = await hp.get();
    expect(q.status, RequestStatus.quoted);
    expect(seen.toString(), 'https://x/api/v1/quote/tok%201');
  });

  test('confirms a selection', () async {
    Map<String, dynamic>? sentBody;
    String? method;
    final mock = MockClient((req) async {
      method = req.method;
      sentBody = jsonDecode(req.body) as Map<String, dynamic>;
      return ok(quoteJson('confirmed', selectedOptionId: 'opt-1'));
    });
    final hp = HealthPayQuoteClient(baseUrl: 'https://x', token: 't', httpClient: mock);
    final q = await hp.confirm('opt-1');
    expect(method, 'POST');
    expect(sentBody, {'optionId': 'opt-1'});
    expect(q.selectedOptionId, 'opt-1');
    expect(q.selectedOption?.providerName, 'Alfa Labs');
  });

  test('maps API error codes to typed errors', () async {
    final hp404 = HealthPayQuoteClient(
        baseUrl: 'https://x', token: 't', httpClient: MockClient((_) async => err('not_found', 404)));
    final hp409 = HealthPayQuoteClient(
        baseUrl: 'https://x', token: 't', httpClient: MockClient((_) async => err('conflict', 409)));
    final hp422 = HealthPayQuoteClient(
        baseUrl: 'https://x', token: 't', httpClient: MockClient((_) async => err('validation_error', 422)));
    await expectLater(hp404.get(), throwsA(isA<NotFoundError>()));
    await expectLater(hp409.confirm('x'), throwsA(isA<ConflictError>()));
    await expectLater(hp422.confirm('x'), throwsA(isA<ValidationError>()));
  });

  test('poll delivers an update and stops at a terminal status', () async {
    var calls = 0;
    final mock = MockClient((_) async {
      calls++;
      return ok(quoteJson('confirmed', selectedOptionId: 'opt-1'));
    });
    final hp = HealthPayQuoteClient(baseUrl: 'https://x', token: 't', httpClient: mock);

    final seen = <RequestStatus>[];
    final stop = hp.poll(
      (q) => seen.add(q.status),
      interval: const Duration(milliseconds: 10),
    );
    // Allow the initial async tick to run.
    await Future<void>.delayed(const Duration(milliseconds: 50));
    stop();

    expect(seen, [RequestStatus.confirmed]);
    expect(calls, 1); // terminal on first poll → no re-schedule
  });
}
